specify how many segments were pruned by each server segment pruner

gortiz opened the pull request:
This PR adds three new metrics that break down the older (and still supported) `numSegmentsPrunedByServer`.

The three metrics are:
- `numSegmentsPrunedInvalid`: Which include both segments that have been pruned because they are actually invalid (aka having docs) or their physical schema doesn't contain all the required columns (quite common when a new column is added but segments were not reloaded)
- `numSegmentsPrunedByLimit`: The ones pruned by SelectionQuerySegmentPruner, for example: `select * from Table limit 10`
- `numSegmentsPrunedByValue`: The ones pruned by ColumnValueSegmentPruner, for example when segments can be pruned by using bloom filters or max/min of each column.

These metrics are shown when the controller is used. It may also be interesting to add a warning in the controller when `numSegmentsPrunedInvalid` is different than 0.

walterddr reviewed (approved):
looks good to me. thanks for adding this support and making it available for users to diagnose invalid segments. 
is it possible to add a test for `SegmentPruner` for this behavior by mocking an invalid segment?

Jackie-Jiang reviewed (commented):
It looks good in general. My only concern is around the empty segment, which IMO shouldn't be count as invalid. Empty segment is normal for certain streaming types e.g. Kinesis. If we emit warning for empty segment, all the query will get the warning, which is undesired

gortiz commented:
> My only concern is around the empty segment, which IMO shouldn't be count as invalid.

Changed

> is it possible to add a test for SegmentPruner for this behavior by mocking an invalid segment?

I've added some tests _inspired_ by the other SegmentPruner tests.

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java:
```diff
@@ -113,6 +113,9 @@ enum MetadataKey {
     SYSTEM_ACTIVITIES_CPU_TIME_NS("systemActivitiesCpuTimeNs", MetadataValueType.LONG),
     RESPONSE_SER_CPU_TIME_NS("responseSerializationCpuTimeNs", MetadataValueType.LONG),
     NUM_SEGMENTS_PRUNED_BY_SERVER("numSegmentsPrunedByServer", MetadataValueType.INT),
+    NUM_SEGMENTS_PRUNED_INVALID("numSegmentsPrunedByInvalid", MetadataValueType.INT),
```
(MAJOR) Please append the new keys to the end. See the javadoc for this enum.
We should associate an id to each key instead of relying on the ordinal of the enum. That is out of the scope of this PR, and we need a newer version data table so that the change is backward compatible

gortiz replied:
Nice catch

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/query/pruner/SegmentPrunerService.java:
```diff
@@ -76,18 +115,40 @@ public List<IndexSegment> prune(List<IndexSegment> segments, QueryContext query)
+    int invalid = 0;
     for (IndexSegment segment : segments) {
-      if (!isInvalidSegment(segment, query)) {
+      boolean isInvalid = isInvalidSegment(segment, query);
```
(minor) We can skip the valid check if a segment is empty
```suggestion
      if (!isEmptySegment(segment)) {
        if (isInvalidSegment(segment, query) {
          invalid++
        } else {
          segments.set(selected++, segment);
        }
      }
```

gortiz reviewed (commented):
Moved ne literals to the end of the enum list
